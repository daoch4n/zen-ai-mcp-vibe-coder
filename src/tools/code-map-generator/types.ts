/**
 * Type definitions for the Code-Map Generator tool.
 * This file contains interfaces for configuration, cache, and other shared types.
 */

/**
 * Main configuration interface for the Code-Map Generator.
 */
export interface CodeMapGeneratorConfig {
  /**
   * Required absolute path to the directory that the code-map generator is allowed to scan.
   * This is a security boundary - the tool will not access files outside this directory.
   */
  allowedMappingDirectory: string;
  
  /**
   * Optional cache configuration for optimizing performance and memory usage.
   */
  cache?: CacheConfig;
  
  /**
   * Optional processing configuration for controlling batch sizes and memory usage.
   */
  processing?: ProcessingConfig;
  
  /**
   * Optional output configuration for controlling where and how output files are saved.
   */
  output?: OutputConfig;
}

/**
 * Configuration for the file-based cache system.
 */
export interface CacheConfig {
  /**
   * Whether the cache is enabled. Default is true.
   */
  enabled: boolean;
  
  /**
   * Maximum number of entries to keep in the cache. Default is 10000.
   */
  maxEntries?: number;
  
  /**
   * Maximum age of cache entries in milliseconds. Default is 24 hours.
   */
  maxAge?: number;
  
  /**
   * Directory where cache files are stored. If not specified, defaults to
   * a '.cache' directory within the code-map-generator output directory.
   */
  cacheDir?: string;
}

/**
 * Configuration for processing options.
 */
export interface ProcessingConfig {
  /**
   * Number of files to process in each batch. Default is 100.
   */
  batchSize?: number;
  
  /**
   * Whether to log memory usage statistics during processing. Default is false.
   */
  logMemoryUsage?: boolean;
  
  /**
   * Maximum memory usage in MB before triggering garbage collection. Default is 1024 (1GB).
   */
  maxMemoryUsage?: number;
}

/**
 * Configuration for output options.
 */
export interface OutputConfig {
  /**
   * Directory where output files are saved. If not specified, defaults to
   * 'code-map-generator' within the VibeCoderOutput directory.
   */
  outputDir?: string;
  
  /**
   * Format for the output. Default is 'markdown'.
   */
  format?: 'markdown' | 'json';
  
  /**
   * Whether to split output into multiple files. Default is true.
   */
  splitOutput?: boolean;
  
  /**
   * Custom prefix for output filenames. Default is timestamp.
   */
  filePrefix?: string;
}

/**
 * Interface for directory structure paths.
 */
export interface DirectoryStructure {
  /**
   * Base output directory (VibeCoderOutput)
   */
  baseOutputDir: string;
  
  /**
   * Tool-specific output directory (VibeCoderOutput/code-map-generator)
   */
  outputDir: string;
  
  /**
   * Cache directory (VibeCoderOutput/code-map-generator/.cache)
   */
  cacheDir: string;
  
  /**
   * File info cache directory (VibeCoderOutput/code-map-generator/.cache/file-info)
   */
  fileInfoCacheDir: string;
  
  /**
   * Metadata cache directory (VibeCoderOutput/code-map-generator/.cache/metadata)
   */
  metadataCacheDir: string;
  
  /**
   * Temporary files directory (VibeCoderOutput/code-map-generator/.cache/temp)
   */
  tempDir: string;
  
  /**
   * Job-specific temporary directory (VibeCoderOutput/code-map-generator/.cache/temp/{jobId})
   */
  jobTempDir: string;
}

/**
 * Interface for path validation result.
 */
export interface PathValidationResult {
  /**
   * Whether the path is valid and within the allowed boundary.
   */
  isValid: boolean;
  
  /**
   * Error message if the path is invalid.
   */
  error?: string;
  
  /**
   * Normalized path if valid.
   */
  normalizedPath?: string;
}
