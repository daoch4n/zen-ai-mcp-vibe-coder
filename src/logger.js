// src/logger.ts
import { pino } from 'pino';
import path from 'path';
import { fileURLToPath } from 'url';
const isDevelopment = process.env.NODE_ENV === 'development';
const effectiveLogLevel = process.env.LOG_LEVEL || (isDevelopment ? 'debug' : 'info');
// --- Calculate paths ---
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
// Log file in the project root directory (one level up from src)
const logFilePath = path.resolve(__dirname, '../server.log');
// --- Create streams ---
// Log to file and also to the original console stream
const streams = [
    { level: effectiveLogLevel, stream: pino.destination(logFilePath) },
    // Redirect console output to stderr when not in development to avoid interfering with MCP stdio
    { level: effectiveLogLevel, stream: isDevelopment ? process.stdout : process.stderr }
];
// Configure the logger
const configuredLogger = pino({
    level: effectiveLogLevel, // Set level here for filtering before transport/multistream
    // --- Add Redaction ---
    redact: {
        paths: [
            'apiKey', // Redact any top-level apiKey
            '*.apiKey', // Redact apiKey in any nested object
            'receivedConfig.apiKey', // Specifically target the observed log structure
            'config.apiKey', // Common config pattern
            'openRouterConfig.apiKey', // Specific object name from index.ts
            'env.OPENROUTER_API_KEY', // If env vars are logged directly
            'env.PERPLEXITY_API_KEY' // Handle other potential keys
            // Add other sensitive keys if necessary, e.g., 'headers.Authorization'
        ],
        censor: '[REDACTED]', // Replace sensitive value with this string
    },
    // --- End Redaction ---
    // Transport is applied *after* multistream, only affects console output here
    transport: isDevelopment
        ? {
            target: 'pino-pretty',
            options: {
                colorize: true,
                translateTime: 'SYS:standard',
                ignore: 'pid,hostname', // Pretty print options
            },
        }
        : undefined, // Use default JSON transport for console when not in development
}, pino.multistream(streams) // Use multistream for output destinations
);
export default configuredLogger;
