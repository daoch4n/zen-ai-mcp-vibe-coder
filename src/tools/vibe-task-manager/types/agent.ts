/**
 * Agent communication and coordination types for the Vibe Task Manager
 */

/**
 * Represents the status of an agent
 */
export type AgentStatus = 'idle' | 'busy' | 'error' | 'offline' | 'initializing';

/**
 * Represents the capabilities of an agent
 */
export type AgentCapability = 
  | 'code_generation' 
  | 'testing' 
  | 'documentation' 
  | 'research' 
  | 'deployment' 
  | 'review' 
  | 'debugging' 
  | 'optimization';

/**
 * Represents an AI agent in the system
 */
export interface Agent {
  /** Unique agent identifier */
  id: string;
  
  /** Human-readable agent name */
  name: string;
  
  /** Agent description */
  description: string;
  
  /** Current status */
  status: AgentStatus;
  
  /** Agent capabilities */
  capabilities: AgentCapability[];
  
  /** Current task assignment */
  currentTask?: string;
  
  /** Task queue */
  taskQueue: string[];
  
  /** Performance metrics */
  performance: {
    tasksCompleted: number;
    averageCompletionTime: number;
    successRate: number;
    lastActiveAt: Date;
  };
  
  /** Configuration */
  config: {
    maxConcurrentTasks: number;
    preferredTaskTypes: string[];
    workingHours?: {
      start: string;
      end: string;
      timezone: string;
    };
  };
  
  /** Communication settings */
  communication: {
    protocol: 'sentinel' | 'direct' | 'webhook';
    endpoint?: string;
    apiKey?: string;
    timeout: number;
  };
  
  /** Metadata */
  metadata: {
    createdAt: Date;
    lastUpdatedAt: Date;
    version: string;
    tags: string[];
  };
}

/**
 * Message types for agent communication
 */
export type MessageType = 
  | 'task_assignment' 
  | 'task_update' 
  | 'task_completion' 
  | 'task_error' 
  | 'status_request' 
  | 'status_response' 
  | 'heartbeat' 
  | 'shutdown' 
  | 'configuration_update';

/**
 * Represents a message in the agent communication protocol
 */
export interface AgentMessage {
  /** Unique message identifier */
  id: string;
  
  /** Message type */
  type: MessageType;
  
  /** Sender agent ID */
  from: string;
  
  /** Recipient agent ID (or 'broadcast' for all agents) */
  to: string;
  
  /** Message payload */
  payload: Record<string, unknown>;
  
  /** Message priority */
  priority: 'low' | 'normal' | 'high' | 'urgent';
  
  /** Timestamp */
  timestamp: Date;
  
  /** Expiration time */
  expiresAt?: Date;
  
  /** Whether this message requires acknowledgment */
  requiresAck: boolean;
  
  /** Correlation ID for request-response patterns */
  correlationId?: string;
  
  /** Retry count */
  retryCount: number;
  
  /** Maximum retries allowed */
  maxRetries: number;
}

/**
 * Agent task assignment information
 */
export interface TaskAssignment {
  /** Assignment ID */
  id: string;
  
  /** Task ID being assigned */
  taskId: string;
  
  /** Agent ID receiving the assignment */
  agentId: string;
  
  /** Assignment timestamp */
  assignedAt: Date;
  
  /** Expected completion time */
  expectedCompletionAt: Date;
  
  /** Assignment priority */
  priority: 'low' | 'normal' | 'high' | 'urgent';
  
  /** Assignment context */
  context: {
    projectId: string;
    epicId: string;
    dependencies: string[];
    resources: string[];
    constraints: string[];
  };
  
  /** Assignment status */
  status: 'pending' | 'accepted' | 'in_progress' | 'completed' | 'failed' | 'cancelled';
  
  /** Progress information */
  progress: {
    percentage: number;
    currentStep: string;
    estimatedTimeRemaining: number;
    lastUpdateAt: Date;
  };
  
  /** Results (when completed) */
  results?: {
    success: boolean;
    output: string;
    artifacts: string[];
    metrics: Record<string, number>;
    logs: string[];
    errors: string[];
  };
}

/**
 * Agent coordination strategy
 */
export interface CoordinationStrategy {
  /** Strategy name */
  name: string;
  
  /** Strategy description */
  description: string;
  
  /** Load balancing algorithm */
  loadBalancing: 'round_robin' | 'least_loaded' | 'capability_based' | 'priority_based';
  
  /** Task distribution rules */
  distributionRules: {
    maxTasksPerAgent: number;
    preferredAgentTypes: Record<string, AgentCapability[]>;
    fallbackStrategy: 'queue' | 'reassign' | 'fail';
  };
  
  /** Conflict resolution */
  conflictResolution: {
    strategy: 'queue' | 'priority' | 'merge' | 'split';
    timeout: number;
    escalation: boolean;
  };
  
  /** Health monitoring */
  healthMonitoring: {
    heartbeatInterval: number;
    timeoutThreshold: number;
    retryAttempts: number;
    failoverEnabled: boolean;
  };
}

/**
 * Sentinel protocol configuration
 */
export interface SentinelProtocolConfig {
  /** Protocol version */
  version: string;
  
  /** Communication settings */
  communication: {
    port: number;
    host: string;
    secure: boolean;
    timeout: number;
    retryInterval: number;
    maxRetries: number;
  };
  
  /** Message settings */
  messaging: {
    maxMessageSize: number;
    compressionEnabled: boolean;
    encryptionEnabled: boolean;
    batchingEnabled: boolean;
    batchSize: number;
  };
  
  /** Agent registration */
  registration: {
    autoRegister: boolean;
    registrationTimeout: number;
    capabilityValidation: boolean;
    authenticationRequired: boolean;
  };
  
  /** Monitoring and logging */
  monitoring: {
    metricsEnabled: boolean;
    loggingLevel: 'debug' | 'info' | 'warn' | 'error';
    performanceTracking: boolean;
    alertingEnabled: boolean;
  };
}

/**
 * Agent orchestration result
 */
export interface OrchestrationResult {
  /** Orchestration session ID */
  sessionId: string;
  
  /** Assigned tasks */
  assignments: TaskAssignment[];
  
  /** Orchestration status */
  status: 'pending' | 'in_progress' | 'completed' | 'failed' | 'cancelled';
  
  /** Overall progress */
  progress: {
    totalTasks: number;
    completedTasks: number;
    failedTasks: number;
    percentage: number;
  };
  
  /** Performance metrics */
  metrics: {
    startTime: Date;
    endTime?: Date;
    totalDuration?: number;
    averageTaskTime: number;
    throughput: number;
    errorRate: number;
  };
  
  /** Issues encountered */
  issues: {
    type: 'agent_unavailable' | 'task_failed' | 'timeout' | 'dependency_error';
    severity: 'low' | 'medium' | 'high' | 'critical';
    description: string;
    affectedTasks: string[];
    resolution?: string;
  }[];
  
  /** Final results */
  results?: {
    success: boolean;
    completedTasks: string[];
    failedTasks: string[];
    artifacts: string[];
    logs: string[];
  };
}
