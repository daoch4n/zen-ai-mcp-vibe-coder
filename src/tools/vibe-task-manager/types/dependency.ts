/**
 * Dependency management types for the Vibe Task Manager
 */

/**
 * Types of dependencies between tasks
 */
export type DependencyType = 'blocks' | 'enables' | 'requires' | 'suggests';

/**
 * Represents a dependency relationship between two tasks
 */
export interface Dependency {
  /** Unique identifier for this dependency */
  id: string;
  
  /** ID of the task that has the dependency */
  fromTaskId: string;
  
  /** ID of the task that is depended upon */
  toTaskId: string;
  
  /** Type of dependency */
  type: DependencyType;
  
  /** Human-readable description of the dependency */
  description: string;
  
  /** Whether this dependency is critical for execution */
  critical: boolean;
  
  /** Metadata */
  metadata: {
    createdAt: Date;
    createdBy: string;
    reason: string;
  };
}

/**
 * Represents a node in the dependency graph
 */
export interface DependencyNode {
  /** Task ID */
  taskId: string;
  
  /** Task title for display */
  title: string;
  
  /** Task status */
  status: string;
  
  /** Estimated hours */
  estimatedHours: number;
  
  /** Priority level */
  priority: string;
  
  /** Dependencies (tasks this node depends on) */
  dependencies: string[];
  
  /** Dependents (tasks that depend on this node) */
  dependents: string[];
  
  /** Depth in the dependency tree */
  depth: number;
  
  /** Whether this node is on the critical path */
  criticalPath: boolean;
}

/**
 * Represents the complete dependency graph for a project
 */
export interface DependencyGraph {
  /** Project ID this graph belongs to */
  projectId: string;
  
  /** All nodes in the graph */
  nodes: Map<string, DependencyNode>;
  
  /** All dependency relationships */
  edges: Dependency[];
  
  /** Topologically sorted task order */
  executionOrder: string[];
  
  /** Critical path through the graph */
  criticalPath: string[];
  
  /** Graph statistics */
  statistics: {
    totalTasks: number;
    totalDependencies: number;
    maxDepth: number;
    cyclicDependencies: string[][];
    orphanedTasks: string[];
  };
  
  /** Graph metadata */
  metadata: {
    generatedAt: Date;
    version: string;
    isValid: boolean;
    validationErrors: string[];
  };
}

/**
 * Configuration for dependency analysis
 */
export interface DependencyAnalysisConfig {
  /** Whether to detect circular dependencies */
  detectCycles: boolean;
  
  /** Whether to optimize the execution order */
  optimizeOrder: boolean;
  
  /** Whether to identify the critical path */
  identifyCriticalPath: boolean;
  
  /** Maximum depth to analyze */
  maxDepth: number;
  
  /** Whether to suggest missing dependencies */
  suggestMissingDependencies: boolean;
  
  /** Whether to validate dependency consistency */
  validateConsistency: boolean;
}

/**
 * Result of dependency analysis
 */
export interface DependencyAnalysisResult {
  /** The analyzed dependency graph */
  graph: DependencyGraph;
  
  /** Issues found during analysis */
  issues: {
    type: 'cycle' | 'orphan' | 'missing' | 'inconsistent';
    severity: 'error' | 'warning' | 'info';
    description: string;
    affectedTasks: string[];
    suggestedFix?: string;
  }[];
  
  /** Optimization suggestions */
  optimizations: {
    type: 'reorder' | 'parallelize' | 'merge' | 'split';
    description: string;
    affectedTasks: string[];
    estimatedImprovement: string;
  }[];
  
  /** Analysis metadata */
  metadata: {
    analyzedAt: Date;
    analysisTime: number;
    configUsed: DependencyAnalysisConfig;
  };
}

/**
 * Dependency validation rules
 */
export interface DependencyValidationRule {
  /** Rule identifier */
  id: string;
  
  /** Rule name */
  name: string;
  
  /** Rule description */
  description: string;
  
  /** Rule severity */
  severity: 'error' | 'warning' | 'info';
  
  /** Validation function */
  validate: (graph: DependencyGraph) => {
    valid: boolean;
    issues: string[];
  };
}

/**
 * Dependency resolution strategy
 */
export interface DependencyResolutionStrategy {
  /** Strategy name */
  name: string;
  
  /** Strategy description */
  description: string;
  
  /** Resolution function */
  resolve: (graph: DependencyGraph) => {
    resolved: boolean;
    modifiedGraph: DependencyGraph;
    changes: string[];
  };
}

/**
 * Dependency graph operations
 */
export interface DependencyGraphOperations {
  /** Add a new dependency */
  addDependency(from: string, to: string, type: DependencyType, description: string): boolean;
  
  /** Remove a dependency */
  removeDependency(dependencyId: string): boolean;
  
  /** Update a dependency */
  updateDependency(dependencyId: string, updates: Partial<Dependency>): boolean;
  
  /** Get all dependencies for a task */
  getDependencies(taskId: string): Dependency[];
  
  /** Get all dependents for a task */
  getDependents(taskId: string): Dependency[];
  
  /** Check if adding a dependency would create a cycle */
  wouldCreateCycle(from: string, to: string): boolean;
  
  /** Get the shortest path between two tasks */
  getShortestPath(from: string, to: string): string[];
  
  /** Get all tasks that can be executed in parallel */
  getParallelExecutableTasks(): string[][];
  
  /** Validate the entire graph */
  validateGraph(): DependencyAnalysisResult;
  
  /** Optimize the execution order */
  optimizeExecutionOrder(): string[];
  
  /** Export graph to various formats */
  exportGraph(format: 'json' | 'yaml' | 'mermaid' | 'dot'): string;
}
