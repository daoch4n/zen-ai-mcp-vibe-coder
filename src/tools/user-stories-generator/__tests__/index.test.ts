// src/tools/user-stories-generator/__tests__/index.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'; // Keep only one import
import { generateUserStories } from '../index.js';
import { OpenRouterConfig } from '../../../types/workflow.js';
import * as researchHelper from '../../../utils/researchHelper.js';
import * as llmHelper from '../../../utils/llmHelper.js'; // Import the new helper
import fs from 'fs-extra';
import { jobManager, JobStatus } from '../../../services/job-manager/index.js'; // Import Job Manager
import { sseNotifier } from '../../../services/sse-notifier/index.js'; // Import SSE Notifier
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js'; // Import CallToolResult
import { ApiError, ToolExecutionError } from '../../../utils/errors.js'; // Import necessary errors
import logger from '../../../logger.js'; // Import logger

// Mock dependencies
vi.mock('../../../utils/researchHelper.js');
vi.mock('../../../utils/llmHelper.js'); // Mock the new helper
vi.mock('fs-extra');
vi.mock('../../../services/job-manager/index.js'); // Mock Job Manager
vi.mock('../../../services/sse-notifier/index.js'); // Mock SSE Notifier
vi.mock('../../../logger.js'); // Mock logger

// Define helper variables for mocks using vi.mocked() for better type handling
const mockPerformResearchQuery = vi.mocked(researchHelper.performResearchQuery);
const mockPerformDirectLlmCall = vi.mocked(llmHelper.performDirectLlmCall); // Mock the new helper
const mockWriteFile = vi.mocked(fs.writeFile);
const mockEnsureDir = vi.mocked(fs.ensureDir);

// Helper to advance timers and allow setImmediate to run
const runAsyncTicks = async (count = 1) => {
  for (let i = 0; i < count; i++) {
    await vi.advanceTimersToNextTimerAsync(); // Allow setImmediate/promises to resolve
  }
};

const mockJobId = 'mock-us-job-id';

describe('User Stories Generator (Async)', () => {
  // Mock data and responses
  const mockConfig: OpenRouterConfig = {
    baseUrl: 'https://api.example.com',
    apiKey: 'test-api-key',
    geminiModel: 'google/gemini-2.5-pro-exp-03-25:free',
    perplexityModel: 'perplexity/sonar-deep-research'
  };
  const mockContext = { sessionId: 'test-session-us' };

  const mockResearchResults = [
    "Mock user personas and stakeholders research data",
    "Mock user workflows and use cases research data",
    "Mock user experience expectations research data"
  ];

  const mockGeneratedUserStories = "# Mock User Stories\n\nThis is a mock user stories document generated by the test.";

  beforeEach(() => {
    // Reset mocks before each test
    vi.clearAllMocks();

    // Default mocks for successful execution
    mockEnsureDir.mockResolvedValue(undefined);
    mockWriteFile.mockResolvedValue(undefined);
    // Setup mock chain for research queries
    mockPerformResearchQuery
        .mockResolvedValueOnce(mockResearchResults[0]) // Personas
        .mockResolvedValueOnce(mockResearchResults[1]) // Workflows
        .mockResolvedValueOnce(mockResearchResults[2]); // Expectations
    mockPerformDirectLlmCall.mockResolvedValue(mockGeneratedUserStories); // Mock the direct call

    // Mock Job Manager methods
    vi.mocked(jobManager.createJob).mockReturnValue(mockJobId);
    vi.mocked(jobManager.updateJobStatus).mockReturnValue(true);
    vi.mocked(jobManager.setJobResult).mockReturnValue(true);

    // Enable fake timers
    vi.useFakeTimers();
  });

   afterEach(() => {
       vi.restoreAllMocks();
       vi.useRealTimers(); // Restore real timers
   });

  it('should return job ID and complete successfully in background', async () => {
    const productDescription = "A test product description";
    const params = { productDescription };

    // --- Initial Call ---
    const initialResult = await generateUserStories(params, mockConfig, mockContext);
    expect(initialResult.isError).toBe(false);
    expect(initialResult.content[0]?.text).toContain(`User stories generation started. Job ID: ${mockJobId}`);
    expect(jobManager.createJob).toHaveBeenCalledWith('generate-user-stories', params);

    // Verify underlying logic not called yet
    expect(mockPerformResearchQuery).not.toHaveBeenCalled();
    expect(mockPerformDirectLlmCall).not.toHaveBeenCalled();
    expect(mockWriteFile).not.toHaveBeenCalled();
    expect(jobManager.setJobResult).not.toHaveBeenCalled();

    // --- Advance Timers ---
    await runAsyncTicks(5); // Allow for research + generation

    // --- Verify Async Operations ---
    // 1. Verify Research
    expect(mockPerformResearchQuery).toHaveBeenCalledTimes(3);
    const researchCalls = vi.mocked(researchHelper.performResearchQuery).mock.calls;
    expect(researchCalls[0][0]).toContain('personas and stakeholders');
    expect(researchCalls[1][0]).toContain('workflows and use cases');
    expect(researchCalls[2][0]).toContain('experience expectations');

    // 2. Verify LLM Call
    expect(mockPerformDirectLlmCall).toHaveBeenCalledTimes(1);
    const llmCallArgs = mockPerformDirectLlmCall.mock.calls[0];
    expect(llmCallArgs[0]).toContain(params.productDescription); // Main prompt check
    expect(llmCallArgs[1]).toContain("# User Stories Generator"); // System prompt check
    expect(llmCallArgs[3]).toBe('user_stories_generation'); // Logical task name check

    // 3. Verify File Saving
    expect(mockWriteFile).toHaveBeenCalledTimes(1);
    const writeFileArgs = mockWriteFile.mock.calls[0];
    expect(writeFileArgs[0]).toMatch(/user-stories-generator[\\/].*a-test-product-description.*-user-stories\.md$/); // Path check
    expect(writeFileArgs[1]).toContain(mockGeneratedUserStories); // Content check

    // 4. Verify Final Job Result
    expect(jobManager.setJobResult).toHaveBeenCalledTimes(1);
    const finalResultArgs = vi.mocked(jobManager.setJobResult).mock.calls[0];
    expect(finalResultArgs[0]).toBe(mockJobId);
    expect(finalResultArgs[1].isError).toBe(false);
    expect(finalResultArgs[1].content[0]?.text).toContain(`User stories generated successfully and saved to:`);
    expect(finalResultArgs[1].content[0]?.text).toContain(mockGeneratedUserStories);

    // 5. Verify SSE Calls (basic)
    expect(sseNotifier.sendProgress).toHaveBeenCalledWith(mockContext.sessionId, mockJobId, JobStatus.RUNNING, expect.stringContaining('Starting user stories generation'));
    expect(sseNotifier.sendProgress).toHaveBeenCalledWith(mockContext.sessionId, mockJobId, JobStatus.RUNNING, expect.stringContaining('Performing pre-generation research'));
    expect(sseNotifier.sendProgress).toHaveBeenCalledWith(mockContext.sessionId, mockJobId, JobStatus.RUNNING, expect.stringContaining('Generating user stories content via LLM'));
    expect(sseNotifier.sendProgress).toHaveBeenCalledWith(mockContext.sessionId, mockJobId, JobStatus.RUNNING, expect.stringContaining('Saving user stories to file'));
  });

  it('should handle research failures gracefully (async)', async () => {
    // Mock a failed research query
    vi.mocked(Promise.allSettled).mockResolvedValueOnce([
      { status: 'rejected', reason: new Error('Research failed') },
      { status: 'fulfilled', value: mockResearchResults[1] },
      { status: 'fulfilled', value: mockResearchResults[2] }
    ]);
    const params = { productDescription: "A test product description" };
    // --- Initial Call ---
    await generateUserStories(params, mockConfig, mockContext);
    // --- Advance Timers ---
    await runAsyncTicks(5);
    // --- Verify Async Operations ---
    expect(mockPerformDirectLlmCall).toHaveBeenCalledTimes(1); // Generation should still run
    const generationPrompt = vi.mocked(llmHelper.performDirectLlmCall).mock.calls[0][0];
    expect(generationPrompt).toContain("### User Personas & Stakeholders:\n*Research on this topic failed.*\n\n"); // Check failure message
    expect(mockWriteFile).toHaveBeenCalledTimes(1); // File should still be saved
    expect(jobManager.setJobResult).toHaveBeenCalledTimes(1); // Job should complete successfully
    expect(vi.mocked(jobManager.setJobResult).mock.calls[0][1].isError).toBe(false);
    expect(sseNotifier.sendProgress).toHaveBeenCalledWith(mockContext.sessionId, mockJobId, JobStatus.RUNNING, expect.stringContaining('Warning: Error during research phase'));
  });

  it('should set job to FAILED if direct LLM call throws error (async)', async () => {
       const llmError = new ApiError("LLM call failed", 500);
       mockPerformDirectLlmCall.mockRejectedValueOnce(llmError); // Throw error
       const params = { productDescription: "Confusing Product" };
       // --- Initial Call ---
       await generateUserStories(params, mockConfig, mockContext);
       // --- Advance Timers ---
       await runAsyncTicks(5);
       // --- Verify Async Operations ---
       expect(mockPerformResearchQuery).toHaveBeenCalledTimes(3); // Research should have run
       expect(mockPerformDirectLlmCall).toHaveBeenCalledTimes(1); // LLM call was attempted
       expect(mockWriteFile).not.toHaveBeenCalled(); // File shouldn't be saved
       expect(jobManager.setJobResult).toHaveBeenCalledTimes(1); // Job should fail
       const finalResultArgs = vi.mocked(jobManager.setJobResult).mock.calls[0];
       expect(finalResultArgs[0]).toBe(mockJobId);
       expect(finalResultArgs[1].isError).toBe(true);
       expect(finalResultArgs[1].content[0]?.text).toContain('Error during background job');
       const errorDetails = finalResultArgs[1].errorDetails as any;
       expect(errorDetails?.message).toContain('Failed to generate user stories: LLM call failed');
   });

   it('should set job to FAILED if file writing fails (async)', async () => {
       const fileWriteError = new Error("Disk full");
       mockWriteFile.mockRejectedValueOnce(fileWriteError);
       const params = { productDescription: "Unsavable Stories" };
       // --- Initial Call ---
       await generateUserStories(params, mockConfig, mockContext);
       // --- Advance Timers ---
       await runAsyncTicks(5);
       // --- Verify Async Operations ---
       expect(mockPerformResearchQuery).toHaveBeenCalledTimes(3);
       expect(mockPerformDirectLlmCall).toHaveBeenCalledTimes(1);
       expect(mockWriteFile).toHaveBeenCalledTimes(1); // Write was attempted
       expect(jobManager.setJobResult).toHaveBeenCalledTimes(1); // Job should fail
       const finalResultArgs = vi.mocked(jobManager.setJobResult).mock.calls[0];
       expect(finalResultArgs[0]).toBe(mockJobId);
       expect(finalResultArgs[1].isError).toBe(true);
       expect(finalResultArgs[1].content[0]?.text).toContain('Error during background job');
       const errorDetails = finalResultArgs[1].errorDetails as any;
       expect(errorDetails?.message).toContain('Failed to generate user stories: Disk full');
  });

  // --- Snapshot Test (Adapted for Async) ---
  it('should set final job result content matching snapshot (async)', async () => {
      const productDescription = "A ticketing application for event management";
      const params = { productDescription };
      const consistentMockStories = `# User Stories: Event Ticketing App\n\n## Epic: User Registration\n### User Story: Account Creation...`;

      // Reset mocks for consistency
      mockPerformResearchQuery.mockReset();
      mockPerformResearchQuery.mockResolvedValue("Consistent mock research.");
      mockPerformDirectLlmCall.mockReset();
      mockPerformDirectLlmCall.mockResolvedValue(consistentMockStories);
      mockWriteFile.mockReset();
      mockWriteFile.mockResolvedValue(undefined);

      // --- Initial Call ---
      await generateUserStories(params, mockConfig, mockContext);
      // --- Advance Timers ---
      await runAsyncTicks(5);
      // --- Verify Async Operations ---
      expect(jobManager.setJobResult).toHaveBeenCalledTimes(1);
      const finalResultArgs = vi.mocked(jobManager.setJobResult).mock.calls[0];
      expect(finalResultArgs[1].isError).toBe(false);
      const finalResult = finalResultArgs[1] as CallToolResult;

      // Snapshot the main content excluding the timestamp and file path
      const resultText = finalResult.content?.[0]?.text ?? '';
      const contentToSnapshot = (resultText as string)
          .replace(/_Generated: .*_$/, '') // Remove timestamp
          .replace(/User stories generated successfully and saved to: .*?-user-stories\.md\n\n/, '') // Remove file path line
          .trim();
      expect(contentToSnapshot).toMatchSnapshot('User Stories Generator Content');

      // Verify file write happened
      expect(mockWriteFile).toHaveBeenCalledTimes(1);
      expect(mockWriteFile).toHaveBeenCalledWith(expect.stringContaining('-user-stories.md'), expect.any(String), 'utf8');
  });
});
