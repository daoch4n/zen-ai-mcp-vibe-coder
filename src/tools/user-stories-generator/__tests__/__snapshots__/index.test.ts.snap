// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Stories Generator (Async) > should set final job result content matching snapshot (async) > User Stories Generator Content 1`] = `
"# User Stories: Event Ticketing App

## Epic: User Registration
### User Story: Account Creation..."
`;
